File: poly_regression.py
"""
    Polynomial Regression
    Author: Sola Gbenro
"""
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
# Matrix of independent variable
# Explicitily make X a matrix of 10 rows 1 column(s), otherwise default is 1d vector
X = dataset.iloc[:, 1:2].values
# DEPENDENT VARIABLE VECTOR (not a matrix, see above)
y = dataset.iloc[:, 2].values

# Splitting the dataset into the Training set and Test set
# no split required, data is very limited

# Feature Scaling
# no feature scaling required, data is very limited

# Fitting Linear REgression to the dataset
lin_reg = LinearRegression()
lin_reg.fit(X, y)

# Fitting Polunomial Regressino to the dataset
# Will transform X into a new matrix of features that will contrain x^2, x^3, x^n
poly_reg = PolynomialFeatures(degree = 3)
X_poly = poly_reg.fit_transform(X)

"""
    Value of X
    array([[ 1],[ 2],[ 3],[ 4],[ 5],[ 6],[ 7],[ 8],[ 9],[10]], dtype=int64)
    
    Value of X_poly
    array([[   1.,    1.,    1.,    1.],
       [   1.,    2.,    4.,    8.],
       [   1.,    3.,    9.,   27.],
       [   1.,    4.,   16.,   64.],
       [   1.,    5.,   25.,  125.],
       [   1.,    6.,   36.,  216.],
       [   1.,    7.,   49.,  343.],
       [   1.,    8.,   64.,  512.],
       [   1.,    9.,   81.,  729.],
       [   1.,   10.,  100., 1000.]])
"""
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, y)

# Visualizing the Linear Regression results
# real observation points
plt.scatter(X, y, color='red')
# predicted salaries 1-10
plt.plot(X, lin_reg.predict(X), color='blue')
plt.title('Truth or Bluff (Linear Regression)')
plt.xlabel('Position Label')
plt.ylabel('Salary')
plt.show()

# Visualizing the Polynomial Regression results
# create a 1d vector from min(X) to max(x) values are [1-10], increment by .1
X_grid = np.arange(min(X), max(X), 0.1)
# plt takes a MATRIX not a VECTOR, will require .reshape()
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color='red')
# predicted salaries 1-10
plt.plot(X_grid, lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color='blue')
plt.title('Truth or Bluff (Polynomial Regression')
plt.xlabel('Position level')
plt.ylabel('Salary')
plt.show()

# Predicting a new result with Linear Regression
# takes a martrix
lin_reg.predict([[6.5]])
# output: array([330378.78787879])


# Predicting a new result with Polynomial Regression
lin_reg_2.predict(poly_reg.fit_transform([[6.5]]))
# output: array([133259.46969697])

File: poly_regression.R
# Polynomial Regression
# Author: Sola gbenro

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
# indexing in R starts at 1 not 0
dataset = dataset[2:3]

# SPlitting the dataset into the Training-set and Test-set 
#install.packages('caTools')
#library(caTools)
#set.seed(123)
# only 10 observations, no need to split
#split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
#trainging_set = subset(dataset, split == TRUE)
#test_set = subset(dataset, split == FALSE)

# Feature scaling
#training_Set = sscale(training_set)
#test_set = scale(test_set)

# Fitting the Linear Regression to the dataset
lin_reg = lm(formula = Salary ~ .,
             data = dataset)
# summary(lin_reg) shows p_value of .00383 for Level right now

"fitting the Polynomial Regression to the dataset
  A poly regression model is a MLR which is composed of 1 independent variable
  and additional independent variables that are polynomial terms of the first independent variable (i.e x^2, x^3, x^n" 
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)

#summary(poly_reg)

# Visualizing the Linear Regression results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Linear Regression') +
  xlab('Level') +
  ylab('Salary')

# visualizing the Polynomial Regression results
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') +
  ggtitle('Truth or Bluff (Polynomial Regression') +
  xlab('Level') +
  ylab('Salary')

# Predicting a new result with Linear Regression
# prediction will be a single value
y_pred = predict(lin_reg, data.frame(Level = 6.5))

# Predicitng a new result with Polynomial Regression
# prediction will be a single value
y_pred_poly = predict(poly_reg, data.frame(Level = 6.5,
                                           Level2 = 6.5^2,
                                           Level3 = 6.5^3,
                                           Level4 = 6.5^4))
